klippy/extras/probe_eddy_current.py

```diff
@@ -301,6 +301,101 @@ def note_probe_and_position(self, start_time, end_time, pos_time):
         self._probe_times.append((start_time, end_time, pos_time, None))
         self._check_samples()
 
+class EddyGatherTapSamples:
+    def __init__(self, printer, sensor_helper, tap_max_decel):
+        self._printer = printer
+        self._sensor_helper = sensor_helper
+        self.tap_max_decel = tap_max_decel
+        # Results storage
+        self._samples = []
+        self._probe_times = []
+        self._probe_results = []
+        self._need_stop = False
+        # Start samples
+        sensor_helper.add_client(self._add_measurement)
+    def _add_measurement(self, msg):
+        if self._need_stop:
+            del self._samples[:]
+            return False
+        self._samples.extend(msg["data"])
+        self._analyze_samples()
+        return True
+    def finish(self):
+        self._need_stop = True
+    def _await_samples(self):
+        # Make sure enough samples have been collected
+        reactor = self._printer.get_reactor()
+        mcu = self._sensor_helper.get_mcu()
+        while self._probe_times:
+            start_time, end_time = self._probe_times[0]
+            systime = reactor.monotonic()
+            est_print_time = mcu.estimated_print_time(systime)
+            if est_print_time > end_time + 1.0:
+                raise self._printer.command_error(
+                    "probe_eddy_current sensor outage")
+            reactor.pause(systime + 0.010)
+    def _lookup_toolhead_pos(self, pos_time):
+        toolhead = self._printer.lookup_object('toolhead')
+        kin = toolhead.get_kinematics()
+        kin_spos = {s.get_name(): s.mcu_to_commanded_position(
+            s.get_past_mcu_position(pos_time))
+            for s in kin.get_steppers()}
+        return kin.calc_position(kin_spos)
+    def _analyze_samples(self):
+        while self._samples and self._probe_times:
+            start_time, end_time = self._probe_times[0]
+            tap_time = []
+            tap_value = []
+            for time, freq, z in self._samples:
+                if start_time <= time < end_time:
+                    tap_time.append(time)
+                    tap_value.append(freq)
+                if time >= end_time:
+                    break
+            while self._samples and self._samples[0][0] < end_time:
+                self._samples.pop(0)
+            if len(tap_time) < 16:
+                raise self._printer.command_error("Not enough samples. "
+                                                  "Trigger too early?")
+            # If samples have gaps this will not work produce adequate data
+            odr = self._sensor_helper.data_rate
```

Comment:
I think it is important that code in one .py file does not peak into the member variables of classes defined in other py files.  Otherwise it becomes really painful to maintain the code.

Also, for what it is worth, I think the code would be easier to understand if we could move the "math manipulation" code into a class separate from the "data gathering" code.  It may also be nice to share one "data gathering" implementation between EddyGatherSamples and EddyGatherTapSamples.